### PR TITLE
Improve TensorRT binding index query

### PR DESCRIPTION
**Pull request by tfeher**
This PR improves the TensorRT binding index query in a way that makes it compatible with optimization profiles. 

Currently profiles are not used in TF-TRT, therefore this PR does not change any existing behavior. It is just a preparatory step to implement dynamic shapes with profiles. 

TensorRT bindings are an array of pointers to input and output buffers for the network.
To [execute TensorRT inference](https://docs.nvidia.com/deeplearning/sdk/tensorrt-api/c_api/classnvinfer1_1_1_i_execution_context.html#a84436f784eb3f0ea9089de2678d77954) we have to specify the bindings.

To specify the pointers for the bindings, we have to query the binding index for each input and output tensor, and fill the data pointer in the bindings array at the correct location specified by the binding index.

Since TensorRT 6, an engine can have a optimization profiles, and each profile has it's own set of [bindings](https://docs.nvidia.com/deeplearning/sdk/tensorrt-api/c_api/classnvinfer1_1_1_i_cuda_engine.html#ac78a69ff4fba78c61c90894b74c0826c). This PR make binding index query compatible with multiple optimization profiles.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 1]
```diff
@@ -544,6 +544,31 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
   }
 }
 
+// Get the binding index of a tensor in an engine.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Gets. Also please add a blank line after this line.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 2]
```diff
@@ -544,6 +544,31 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
...
+// The binding index is looked up using the tensor's name and the profile idx.
+// Profile idx should be set to zero, if we do not have optimization profiles.
+Status GetTrtBindingIndex(const char* tensor_name, int profile_idx,
+                          const nvinfer1::ICudaEngine* cuda_engine,
+                          int* binding_idx) {
+  *binding_idx = cuda_engine->getBindingIndex(tensor_name);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Can we add this comment?
// This is binding_index_within_profile.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 3]
```diff
@@ -544,6 +544,31 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
...
+    LOG(ERROR) << msg;
+    return errors::NotFound(msg);
+  }
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+  int n_profiles = cuda_engine->getNbOptimizationProfiles();
+#else
+  int n_profiles = 1;
+#endif
+  // If we have more then one optimization profiles then the binding idx
+  // depends on the profile number
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Can we replace this comment with this:
binding_index_within_engine = binding_index_within_profile + profile_index * bindings_per_profile

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 4]
```diff
@@ -569,19 +594,17 @@ bool TRTEngineOp::ExecuteTrtEngine(OpKernelContext* ctx,
...
   // Setup engine inputs.
   for (int i = 0; i < ctx->num_inputs(); i++) {
     const string input_name = StrCat(IONamePrefixes::kInputPHName, i);
-    const int binding_index = cuda_engine->getBindingIndex(input_name.c_str());
-    if (binding_index == -1) {
-      const string msg =
-          StrCat("Input node ", input_name, " not found, at ", name());
-      LOG(ERROR) << msg;
-      ctx->SetStatus(errors::NotFound(msg));
+    int binding_index = -1;
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Do we have to initialize binding_index here?

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 5]
```diff
@@ -544,6 +544,31 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
   }
 }
 
+// Get the binding index of a tensor in an engine.
+// The binding index is looked up using the tensor's name and the profile idx.
+// Profile idx should be set to zero, if we do not have optimization profiles.
+Status GetTrtBindingIndex(const char* tensor_name, int profile_idx,
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Let's replace 'idx' with index so that we use index consistently.

**Review by bixia1 — COMMENTED** [review]
Looks good to me. Would you please remove the use of "PR #36379" in this PR description as it is not important.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc (hunk 3 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: Added this comment.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc (hunk 4 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc]: No, it is not absolutely necessary.  I have removed the initialization.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review. I have addressed the issues.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**